### PR TITLE
Time formatter

### PR DESCRIPTION
**Pull request by kkimble006**
### Added

- 

### Changed

- 

### Deprecated

- 

### Removed

- 

### Fixed

- 

### Security

- 

## Testing

---

[In Atlantis we use Github's built in pull request reviews](https://help.github.com/en/articles/about-pull-request-reviews).

Random photo of Atlantis

### CODE REVIEW

File: packages/components/src/TimeFormatter/TimeFormatter.tsx [hunk 1]
```diff
@@ -0,0 +1,43 @@
+import React from "react";
+
+interface TimeFormatterProps {
+  /**
+   * Civil Time string to be formatted
+   * 17:45:00.000000000
+   * 11:30:00.000000000
+   */
+  readonly civilTime: string;
```
Comment (JordanLos) [inline packages/components/src/TimeFormatter/TimeFormatter.tsx]: @kkimble006 Maybe not directly related to the work in your PR. I'm just curious why we're using `readonly` for props? I think I've seen it in other places too, but I'm pretty sure typescript makes props immutable (i.e. readonly) by default. 

Just to try it out, you can remove the `readonly` keyword, attempt to change the value using something like `props.civilTime = 'someOtherString'` and it will return a `TypeError`.

I'm trying to find some documentation that _guarantees_  props are readonly to confirm. Some starters:
https://reactjs.org/docs/components-and-props.html#props-are-read-only

image

@seanhealy @darryltec  Maybe you could also add some insight here?

File: packages/components/src/TimeFormatter/TimeFormatter.tsx [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+export function TimeFormatter(props: TimeFormatterProps) {
+  const timeSplit = props.civilTime.split(":");
+  const hour = parseInt(timeSplit[0]);
+  const minute = parseInt(timeSplit[1]);
+
+  const amOrPm = getAmPmLabel(hour, props.useMilitaryTime);
+  const formattedHour = setCorrectHourFormat(hour, props.useMilitaryTime);
+
+  return `${formattedHour}:${minute} ${amOrPm}`;
```
Comment (JordanLos) [inline packages/components/src/TimeFormatter/TimeFormatter.tsx]: @kkimble006 Do you think there's any value in error checking any of these values (e.g. hours not between 0-23 and minutes/seconds not between 0-59)? Or should we just trust that whoever is consuming this component does so responsibly?

File: packages/components/src/TimeFormatter/TimeFormatter.tsx [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+function getAmPmLabel(hour: number, useMilitaryTime: boolean) {
+  if (useMilitaryTime === true) {
+    return "";
+  }
+
+  return hour >= 12 ? "PM" : "AM";
+}
+
+function setCorrectHourFormat(hour: number, useMilitaryTime: boolean) {
```
Comment (JordanLos) [inline packages/components/src/TimeFormatter/TimeFormatter.tsx]: @kkimble006 I noticed an edge case here: If you use `useMilitaryTime={false}` but set the hour to `0`, which is a valid hour in military time, it should return `12:xx AM`, but instead return `0:xx AM`, which is not a valid civilian time. 
image

File: packages/components/src/TimeFormatter/TimeFormatter.mdx [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+import { Playground, Props } from "docz";
+import { ComponentStatus } from "@jobber/docx";
+import { TimeFormatter } from ".";
+
+# TimeFormatter
+
+<ComponentStatus stage="pre" responsive="no" accessible="no" />
+
+In Jobber a TimeFormatter is used to …
```
Comment (JordanLos) [inline packages/components/src/TimeFormatter/TimeFormatter.mdx]: @kkimble006 Did you want to leave a quick description of the component here?

File: packages/components/src/TimeFormatter/TimeFormatter.test.tsx [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import React from "react";
+import renderer from "react-test-renderer";
+import { cleanup } from "@testing-library/react";
+import { TimeFormatter } from ".";
+
+afterEach(cleanup);
+
+it("renders a TimeFormatter", () => {
```
Comment (JordanLos) [inline packages/components/src/TimeFormatter/TimeFormatter.test.tsx]: @kkimble006 Time and dates often have edge cases and are prone to simple mistakes, which mean they could benefit from unit tests. What do you think about adding some simpler ones now? I could think of:
1) Military Time displays properly
2) Civilian time displays properly
3) `12` AM displays properly in civilian and military.

File: packages/components/src/TimeFormatter/TimeFormatter.mdx [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+## Simple Example
+
+<Playground>
+  <TimeFormatter civilTime={"11:30:00.000000000"} useMilitaryTime={false} />
```
Comment (seanhealy) [inline packages/components/src/TimeFormatter/TimeFormatter.mdx]: I'm not sure that `MilitaryTime` is a broad name to use here.

Comment (seanhealy) [reply packages/components/src/TimeFormatter/TimeFormatter.mdx]: Perhaps `useAnteMeridiem` or use `use12HourClock`? I feel like there is a more conventional name here I'm just not thinking of it…

File: packages/components/src/TimeFormatter/TimeFormatter.tsx (hunk 1 above)
Comment (seanhealy) [reply packages/components/src/TimeFormatter/TimeFormatter.tsx]: The readonly here is telling typescript that the property is readonly. This allows the compiler to find this error for us instead of relying on the codepath being executed and reported at runtime.

image

vs

image

It is nice to know that react is double checking this for us though too. :)

File: packages/components/src/TimeFormatter/TimeFormatter.tsx (hunk 2 above)
Comment (seanhealy) [reply packages/components/src/TimeFormatter/TimeFormatter.tsx]: I feel like since this component is accepting a CivilTime we should rely on it to ensure that it is a valid time…

Comment (seanhealy) [reply packages/components/src/TimeFormatter/TimeFormatter.tsx]: We are probably also best using the browsers built in locale formatters here. They are going to deal with a lot more edge cases so we don't have to. :)

eg
```ts
(new Date('August 19, 1975 23:15:00 GMT+00:00')).toLocaleTimeString(undefined, {timeStyle: "short"})
```

File: packages/components/src/TimeFormatter/TimeFormatter.tsx [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+  /**
+   * Decide to show 24 hour vs.
+   * 12 hour display
+   */
+  readonly useMilitaryTime: boolean;
+}
+
+export function TimeFormatter(props: TimeFormatterProps) {
+  const timeSplit = props.civilTime.split(":");
```
Comment (seanhealy) [inline packages/components/src/TimeFormatter/TimeFormatter.tsx]: civilTime here should be an instance of `CivilTime`. See:

https://github.com/GetJobber/atlantis/blob/master/packages/components/src/InputTime/InputTime.tsx#L2

https://github.com/tc39/proposal-temporal/blob/master/objects.md#civiltime-

File: packages/components/src/TimeFormatter/TimeFormatter.tsx (hunk 3 above)
Comment (kkimble006) [reply packages/components/src/TimeFormatter/TimeFormatter.tsx]: Good Catch!

File: packages/components/src/TimeFormatter/TimeFormatter.test.tsx (hunk 5 above)
Comment (kkimble006) [reply packages/components/src/TimeFormatter/TimeFormatter.test.tsx]: We are not using https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toLocaleTimeString to set the time, based on a CivilTime object. I don't think we need to test this.

File: packages/components/src/FormatTime/FormatTime.mdx [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+  <FormatTime time={new CivilTime(18, 35)} use24HourClock={true} />
+  <br />
+  <FormatTime time={new CivilTime(2, 35)} use24HourClock={false} />
+  -
+  <FormatTime time={new CivilTime(18, 35)} use24HourClock={false} />
+</Playground>
+
+## FormatTime Properties
+
+<Props of={FormatTime} />
```
Comment (kkimble006) [inline packages/components/src/FormatTime/FormatTime.mdx]: Just two comments, 
1.) This `Props of`, is not actually listing anything, this section is empty when I view it (I re-ran npm start) and refreshed. 
2.) I think the package-lock changes were from me, I am not sure if they are needed?? Should we remove them before merging?

Comment (seanhealy) [reply packages/components/src/FormatTime/FormatTime.mdx]: 1. They are showing here: https://deploy-preview-87--jobber-atlantis.netlify.com/components/time-formatter

image

Sometimes when running in dev mode on your local machine you need to do an `npm run clean` to clear some caching. The Netlify branch previews are usually an easier way to test things though. 😄 

2. The lock file should be fine.
